### PR TITLE
fix(multiprocessing): Honor join() timeout even if processing pool is overloaded [INC-378]

### PR DESCRIPTION
**Pull request by untitaker**
`Pool.join` will block for more than `timeout` seconds if there is a task that takes more than `timeout` seconds. We can just terminate the pool though, because we have effectively reimplemented `join()` ourselves using `__check_for_results`.

Note that when I say "task", I mean "task" from the POV of `multiprocessing`, which translates to a call to `apply_async`, which is an entire message batch from arroyo POV

This effect is potentially made worse if `next_step` is rejecting all messages:

* When running `git checkout origin/main arroyo/`, the new tests hang forever in join, because the batch keeps getting reprocessed
* When additinally removing `side_effect = MessageRejected()` from the new tests, they fail because `join()` "only" terminates whenever the pool is drained, which is 10 seconds.

The fact that we reprocess the remaining batch on backpressure is kinda dumb anyway, let's fix it in #238

### CODE REVIEW

File: arroyo/processing/strategies/run_task_with_multiprocessing.py [hunk 1]
```diff
@@ -601,23 +610,14 @@ def join(self, timeout: Optional[float] = None) -> None:
...
-            except multiprocessing.TimeoutError:
-                pass
+        self.__check_for_results(
+            timeout=timeout,
+        )
 
         self.__pool.close()
 
         logger.debug("Waiting for %s...", self.__pool)
-        # ``Pool.join`` doesn't accept a timeout (?!) but this really shouldn't
```
Comment (onewland) [inline arroyo/processing/strategies/run_task_with_multiprocessing.py:617]: "shouldn't"

File: arroyo/processing/strategies/run_task_with_multiprocessing.py [hunk 2]
```diff
@@ -601,23 +610,14 @@ def join(self, timeout: Optional[float] = None) -> None:
...
-        # block for any significant amount of time unless something really went
-        # wrong (i.e. we lost track of a task)
-        self.__pool.join()
+        self.__pool.terminate()
```
Comment (lynnagara) [inline arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: Is this change actually needed now that RunTaskWithMultiprocessing.join() is applying the proper timeout in the check for results step? I think it's better to shut down cleanly with `pool.close()` + `pool.join()` rather than `terminate()`.

Comment (untitaker) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: if you do that and submit a task that takes 10 seconds, and call `join(timeout=3)`, `__check_for_results` will return after 3 seconds. then, `pool.join()` will block for another 7 seconds because the task is still not done

Comment (lynnagara) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: Ok. Is there a reason for keeping the `pool.close()` call directly before this?

Comment (untitaker) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: Ah, no. I also wonder if we should block on the pool shutting down by calling terminate + join instead of just terminate. I am not sure we should really block rebalance on it

Comment (lynnagara) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: `terminate()` makes sense to me if we have exceeded the deadline already. Though if no timeout is passed, I think it's more correct to honor that and close + join instead of terminate.

Comment (untitaker) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: by the time we call `terminate()`, the deadline is either guaranteed to have passed, or the pool is guaranteed to be empty.

If no deadline is set, `__check_for_results` will block until the pool is empty. So if we `join` after that, it can only either block for (much) longer than the deadline allows, or do nothing useful at all

Comment (lynnagara) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: Ah I see, terminate makes sense then. Still, why do we call `pool.close()` before `pool.terminate()`?

**Review by lynnagara — APPROVED** [review]

File: arroyo/processing/strategies/run_task_with_multiprocessing.py (hunk 2 above)
Comment (untitaker) [reply arroyo/processing/strategies/run_task_with_multiprocessing.py:620]: Oh yeah I forgot about that, will fix on Monday
